### ISSUE TITLE
st terminal needs to change <delete> to use <delete>

### ISSUE THREAD

**Issue by pickfire** [fish-shell/fish-shell#2139]
Hi,
  Fish shell (friendly interactive shell) is design to be friendly. [`st`](st.suckless.org) required to do something in some programs in order to use delete. Look http://git.suckless.org/st/tree/FAQ#n32 for more info. I hope fish shell can avoid these issues to become friendly to the users and [prevent unneeded configuration](http://fishshell.com/docs/current/design.html#conf).

**Comment by zanchey** [comment]
I'm struggling to understand. Can you describe what the difference between you are seeing and what you expect is?

**Comment by pickfire** [comment]
Without `tput smkx` or `printf '\033[?1h\033=' >/dev/tty` in config, when you press <**DELETE**> in st:
```
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
➜  ~ [P
```

**Comment by zanchey** [comment]
Useful reading:
 * http://invisible-island.net/xterm/xterm.faq.html#xterm_arrows
 * http://zsh.sourceforge.net/FAQ/zshfaq03.html#l25
*  http://stackoverflow.com/questions/13585131/keyboard-transmit-mode-in-vt100-terminal-emulator

AIUI:
 * the terminal keyboard can be in keypad_local/normal/raw/non-transmit mode or keypad_xmit/application/transmit mode (all of these names are used)
 * by default the terminal assumes that the keypad is in "normal" mode
 * in "normal" mode stterm does not send any codes for the keypad delete key, sends an ASCII DEL for the standard delete key, and sends `ESC [ D` for the keypad left and standard left keys
 * in "application" mode stterm does send the delete key code listed in its termcap under `kdch` for both the keypad and the standard delete key, and sends `ESC O t` from the keypad left and `ESC O D` from the standard left key

This is different to, say, gnome-terminal, which always sends the key code listed under `kdch` for the standard delete and keypad delete keys regardless of keypad mode. Setting "application mode" produces `ESC O D` from both left keys and "normal mode" sends `ESC [ D` for both left keys.

I'm not terribly convinced that stterm actually does sensible things here but I think it would be reasonable to do the same thing that ZSH is configured to do under Debian: send `smkx` before entering the prompt and `rmkx` when leaving. Adding the extra bindings that stterm sends is an additional exercise.

**Comment by pickfire** [comment]
I don't understand what you mean? Do you mean to set `smkx` manually in `fish.config`?

**Comment by zanchey** [comment]
No, to echo the smkx sequence when going accepting input and rmkx when leaving that state.

**Comment by pickfire** [comment]
On Wed, Jul 01, 2015 at 11:40:35PM -0700, David Adam wrote:
>No, to echo the smkx sequence when going accepting input and rmkx when leaving that state.

In fish? Manually?

-- 
 _____________________________________
< Do what you like, like what you do. >
 -------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

**Comment by zanchey** [comment]
In fish.

**Comment by pickfire** [comment]
Manually?

**Comment by pickfire** [comment]
Do you think I need to patch `st` for the *smkx*?

**Comment by faho** [comment]
Putting this in config.fish should solve the issue:

```fish
switch $TERM
	case 'st-*' # suckless' simple terminal
                # Enable keypad, do it once before fish_postexec ever fires
		tput smkx
		function st_smkx --on-event fish_postexec
			tput smkx
		end
		function st_rmkx --on-event fish_preexec
			tput rmkx
		end
end
```

Should we ship this by default, e.g. in __fish_config_interactive?

>Do you think I need to patch st for the smkx?

That would work too, though I don't think st upstream would take it.

**Comment by pickfire** [comment]
I said that `k` doesn't work, https://github.com/fish-shell/fish-shell/issues/2309#issuecomment-136667386 talked about this.

> That would work too, though I don't think st upstream would take it.

We could still add that to http://st.suckless.org/patches/ first and see if they want to take that.

In a nutshell, the problem is with `smkx` or `delete`?

**Comment by faho** [comment]
>I said that k doesn't work, #2309 (comment) talked about this.

There you said it didn't work in bash, and then you later managed to do it anyway.

>We could still add that to http://st.suckless.org/patches/ first and see if they want to take that.

It's a major change so I don't think they'd appreciate it. But if you want, go ahead.

>In a nutshell, the problem is with smkx or delete?

No. The problem is st seems to be the only terminal that sends weird keycodes when the keypad is not explicitly activated. It's not "delete" since it also affects other keys. It's not "smkx" but the fact that it's necessary.

**Comment by pickfire** [comment]
> There you said it didn't work in bash, and then you later managed to do it anyway.

Sorry, I only show you the link for the solution for **bash**, that doesn't happened in **fish**.

> It's a major change so I don't think they'd appreciate it. But if you want, go ahead.

Yeah, I can do that, I just don't know how to check which keys is abnormal.

> No. The problem is st seems to be the only terminal that sends weird keycodes when the keypad is not explicitly activated. It's not "delete" since it also affects other keys. It's not "smkx" but the fact that it's necessary.

So, **st** is wrong, **fish** is correct is what you want to say? (took from st)

But it sems that `zsh` is doing everything correct, I haven't found an issue with `zsh` except **speed issues**.

**Comment by faho** [comment]
>So, st is wrong, fish is **right* is al what you want to say?

st does something apparently nobody else does, but I guess they'd like to keep their way and they might have a reason for it.

>But it sems that zsh is doing everything correct

It's quite possible zsh already has a hack like what I proposed above.

**Comment by faho** [comment]
Since it looks like the st people aren't going to budge, we have three possible courses of action:

- Not support st and recommend users use something else

- Do the "smkx" dance in c(++)

- Do the dance in fishscript

The first is of course the simplest but also basically hostile, the second is possible but I prefer to put as much as possible into fishscript so it can be modified more easily.

To test the third approach I have put the following into my config.fish:

```fish
tput smkx ^/dev/null
function fish_enable_keypad_transmit --on-event fish_postexec
    tput smkx ^/dev/null
end

function fish_disable_keypad_transmit --on-event fish_preexec
    tput rmkx ^/dev/null
end
```

and tested the delete key in st, konsole, tmux, roxterm, pantheon-terminal, rxvt-unicode and gnome-terminal. It made st "just work" and didn't break anything anywhere else. This is also basically the same thing Debian's /etc/zsh/zshrc does, also without conditition.

If anyone could test it in the OSX terminals (iTerm2 et al), I'd add it to __fish_config_interactive.

**Comment by pickfire** [comment]
> Do the "smkx" dance in c(++)

I like that. Let's dance. Is there anything I can help, I only knew some C. I will take a look in it.

I like `st`, it is fast unlike other terminal and doesn't break any application I use except fish so I think I won't be hostile.

**Comment by faho** [comment]
As mentioned on the suckless mailing list, even xterm _reacts_ to smkx/rmkx. If it is sent smkx, the home key becomes "^[OH" instead of "^[[H" (which should be "\e\[H"), so we might be able to remove some of the manual escape bindings.

**Comment by pickfire** [comment]
Okay, but what about the delete key?

**Comment by faho** [comment]
That fishscript solution doesn't quite work because neither fish_postexec nor fish_prompt events fire when a command is aborted (e.g. via ctrl-c). Aborting `sleep 5` will lead to rmkx being set because `tput smkx` won't be executed.

That should probably be fixed either way, but for now I'm testing if disabling the keypad (rmkx) is actually necessary.

**Comment by pickfire** [comment]
What is the difference between `smkx` and `rmkx`?

**Comment by faho** [comment]
`man terminfo` - "smkx" is "enter keyboard_transmit mode", "rmkx" is "leave keyboard_transmit mode".

What we want is keyboard_transmit mode because only then do some terminals actually send the sequences they specify in terminfo. It's annoying and a relic of VT220, but that's the way it is, and quite possibly the way it always will be.

**Comment by zanchey** [comment]
Is there any way of querying the current mode? I couldn't see one, but then I don't even know if it makes sense to set it back for child processes.

**Comment by zanchey** [comment]
My temptation would be to just have fish emit `smkx` on every entry into interactive mode. I do actually use a VT520 very occasionally, so I could probably test that out at some stage.

**Comment by faho** [comment]
>Is there any way of querying the current mode? I couldn't see one

Me neither, but I find this stuff basically ungoogleable. I can't even find a reason this is a thing other than "that's how VT220s did it". I mean, why would you possibly want a mode that sends sequences that are, as far as I can tell, not documented anywhere and term-specific?

>I don't even know if it makes sense to set it back for child processes.

The best reason I found for doing that is "that's what zsh recommends", and what Debian actually has in the default /etc/zsh/zshrc.

**Comment by pickfire** [comment]
Maybe there is a difference between it that you can find something to differentiate between them.

**Comment by zanchey** [comment]
Revisiting this, the [Xterm FAQ page](http://invisible-island.net/xterm/xterm.faq.html#xterm_arrows) says:

> Terminfo entries are written assuming that the application has initialized the terminal using the smkx string before it is able to match the codes given for the cursor or keypad keys.

I think fish should do that.

**Comment by pickfire** [comment]
Yeah, I think so. So should we do it in `c++`?

**Comment by zanchey** [comment]
See #2371 for a PR.

**Comment by faho** [comment]
Confirmed working in st. Thanks @zanchey!

**Comment by faho** [comment]
Should we now check if there's escape sequences in fish_{default,vi}_key_bindings we can delete?

**Comment by zanchey** [comment]
I'm always a bit wary, because there's so many terminals that identify as "xterm" with different behaviour, and the existing bindings probably aren't hurting anyone.

`git blame` might be helpful to work out if there's specific terminals that require certain bindings, but I'd leave both `keypad_local` and `keypad_xmit` mode bindings in.

**Comment by zanchey** [comment]
I'm going to have to back this out for now due to #2406.

**Comment by pickfire** [comment]
@zanchey Any idea how to fix it? I just figured out that this issue is reopened recently when <delete> fails in st. What about adding a `TERM` check in c++?

**Comment by faho** [comment]
Retargeting this as I don't think we'll get to this for 2.3

**Comment by krader1961** [comment]
The core problem is that terminfo only has sequences for `keypad_xmit` (`smkx`) and `keypad_local` (`rmkx`). Most of the xterm (semi-)compatible terminal definitions modify both cursor and keypad mode in those sequences rather than just keypad mode. Ideally there would be a terminfo sequence fish could send to enable just cursor application mode. That's because the `key_left` and related sequences typically assume cursor application mode has been enabled.

Presumably programs like vim either have custom (i.e., not found in terminfo) mappings to deal with keypad application mode being enabled by `smkx` or users have to appropriately configure their terminal. When I started typing `vim keypad` into Google it autocompleted my search with `vim keypad doesn't work`. The results of that search is an almost even split between configuring the terminal to disable keypad application mode and add custom mappings to vim such as in [this StackExchange thread](http://apple.stackexchange.com/questions/201337/number-pad-does-not-work-in-vim-and-less).

For more info see section 4.4.4 of this [PuTTY documentation](https://the.earth.li/~sgtatham/putty/0.60/htmldoc/Chapter4.html#config-appcursor). And this [StackOverflow](http://stackoverflow.com/questions/21175561/terminfo-smkx-and-application-cursor-keys-vs-application-keypad) thread.

I believe we should reinstate @zanchey's change to send the smkx/rmkx sequences, update our FAQ with relevant information, and add custom bindings to deal with keypad application mode. Using iTerm2 on OS X sending smkx works for me and allows `fish_key_reader` to report the correct key name for the cursor keys. But note that iTerm2 has "allow application keypad mode" not selected by default so I don't need

**Comment by krader1961** [comment]
Also, this [Wikia page](http://vim.wikia.com/wiki/PuTTY_numeric_keypad_mappings) regarding PuTTY and the numeric keypad.

**Comment by pickfire** [comment]
@krader1961 So does it works now?

**Comment by faho** [comment]
@pickfire: There's not been a new commit.

@krader1961: To me it seems like we can choose between scylla and charybdis - either we don't send smkx/rmkx and some terminals send weird sequences that we need to hardcode, or we send them and some terminals send weird sequences we need to hardcode. Depends on which gets us fewer weird sequences, I guess. From the data we have, it's suckless' st and what you are using vs gnome-terminal and putty?

It might also be nice to hardcode a key->sequence mapping, not just bind the sequences, so we and users can just use `bind -k`.

**Comment by krader1961** [comment]
@pickfire: Sorry for the confusion. I was pointing out that @zanchey's change works for me using iTerm2 on OS X and in theory should work for anyone who disables the stupid "keypad application mode" of their terminal.

@faho: Both solutions have problems. The advantage of sending smkx is that the sequences sent by the arrow keys will match the terminfo definition. I definitely like the idea of supporting defining additional key to sequence mappings so that we could use symbolic key names in the `bind` statements . Tmux, for example, lets you do something similar. These lines are from my *.tmux.conf* to tell it my terminal supports true-colors by inserting a "Tc" terminfo attribute:

```
set-option -ga terminal-overrides ",xterm-256color:Tc"
set-option -ga terminal-overrides ",screen-256color:Tc"
```

But we should take a discussion of how that would work to a new issue.

**Comment by floam** [comment]
The way we should do overrides is at compile time,  not a user configurable thing. I think we are doing something to generate C for TERM=ansi, right? There are references to a fallback TERM in our error output, at least. `tput smkx` won't work if your tput expects termcap names.

Since suckless sets TERM well, it's easy to do a good fix. The fix might just be to grab ncurses-6.0-20160813's terminfo.src and use:

```
 infocmp -C -e -x -T st
	{
		st_alias_data,
		(char *)0,	/* pointer to string table */
		st_bool_data,
		st_number_data,
		st_string_data,
#if NCURSES_XNAMES
		(char *)0,	/* pointer to extended string table */
		(char **)0,	/* ...corresponding names */
		44,		/* count total Booleans */
		39,		/* count total Numbers */
		414,		/* count total Strings */
		0,		/* count extensions to Booleans */
		0,		/* count extensions to Numbers */
		0,		/* count extensions to Strings */
#endif /* NCURSES_XNAMES */
	}
```
and
```
infocmp -C -E -x -T st
```

static char st_alias_data[] = "st|stterm| simpleterm 0.4.1";

static char st_s_cbt            [] = "\033[Z";
static char st_s_bel            [] = "\007";
static char st_s_cr             [] = "\015";
static char st_s_csr            [] = "\033[%i%p1%d;%p2%dr";
static char st_s_tbc            [] = "\033[3g";
static char st_s_clear          [] = "\033[H\033[2J";
static char st_s_el             [] = "\033[K";
static char st_s_ed             [] = "\033[J";
static char st_s_hpa            [] = "\033[%i%p1%dG";
static char st_s_cup            [] = "\033[%i%p1%d;%p2%dH";
static char st_s_cud1           [] = "\012";
static char st_s_home           [] = "\033[H";
static char st_s_civis          [] = "\033[?25l";
static char st_s_cub1           [] = "\010";
```
To burn in a "st" definition.

**Comment by floam** [comment]
You literally just #Include it and "compile it in".

**Comment by faho** [comment]
>It might from a totally different curses - we don't want to use it.

Yes, we do. We'll use whatever curses fish is compiled against. If that has a problem, then that's an issue there.

That's basic dynamic linking for you.

(Note that I'm not saying we should call the `tput` command)

**Comment by floam** [comment]
>> It might from a totally different curses - we don't want to use it.
>Yes, we do.

Well, I don't want to call run the `tput` program - we'd need to parse the output from it and other tools like `infocmp` to figure out if it worked or not. 

We can't use `keypad_xmit` there, it has to be `smkx`... or sometimes `ke`, because it might come from a wildly different planet than us where they don't even do `terminfo` - curses can be good in this case, but `tput` signatures not so much. And at the end of the day anybody shelling out to `tput` is not getting built with our fallback TERMs we might compile in.

`tparm()` is a lot more pleasant. It's easier to keep track of the global state, as well, and it's not torture slow. 

> (Note that I'm not saying we should call the tput command)
Oh cool.

**Comment by floam** [comment]
The problem with tput using a different library is just the concern, especially with an obscure terminal, it's more and more likely it's using a different cookbook for the same TERM=foo for us.

There's only a tput utility, no tput function we can call dynamically from C. 

We don't need to actually hack the terminfo entries, of course. It's an extreme measure.

**Comment by faho** [comment]
>tparm() is a lot more pleasant. It's easier to keep track of the global state, as well, and it's not torture slow. 

Okay. Sorry, I thought you meant we should compile-in our own terminfo and use that.

No, of course using the ncurses provided functions is the correct solution here. Ideally we'd then always get every term to send the sequence its terminfo (or tcap or whatever) defines, and can remove all the hardcoding from our bindings and make some more terminals happy.

**Comment by floam** [comment]
>Ideally we'd then always get every term to send the sequence its terminfo (or tcap or whatever) defines, and can remove all the hardcoding from our bindings and make some more terminals happy.

Yeah, actually the symbol names for the basic capabilities in term. h never change - whether it is terminfo or termcap `set_a_foreground` is the symbol name in the API that makes their curses do whatever it is is does for that - in our C there is never any worries about that part. It's only when you're doing the compiling of terminfo/termcap entries or having a user toss "setaf' or "AF" around that the difference is the problem.
* except bind -k uses them: that's basically OK here though.

Everything is just peachy until we have a bunch of users with some TERM that is missing a 512-color savvy (whatever) `set_a_foreground` for their new terminal using TERM=xterm-512color: what we've done to date for these kinds of problems is this when we can't get a workable string from tparm:
    + Not notice that it just used the old escape format for less colors after returning an error code - the text maybe goes black or disappears or rings the terminal bell or is white.
    + We decide it probably will work anyhow to emit the escape xterm on our box emits: we printf() that.
    + Or we totally give up and set TERM=ansi hoping for the best.

**Comment by floam** [comment]
The kind of "hack" like described above can be better than what we do currently, though. 

Back to reality: the trouble has sometimes been the user has TERM=xterm but we think their xterm can do 256 colors or they won't hurt. If we were to make an idealized xterm-256color entry and then convert it to C and compile it in as fishunknown-256color, what we could at least do is switch to *that* at runtime internally when things are looking grim, and then let that run the show form there. We can at least, this way, **not do the nasty** stuff later in the code where we spin plates on one foot trying to pull off a big lie. We can really just get a string from `tparm(set_a_foreground, 256fgval))` check for error values returned, and write that to the screen if it worked or not rite it if it did not, and know that we sent the right thing to everyone we can, instead of nesting 5 levels deep with if/else clauses.

It can solve everything as long as we put out releases and update the entries time to time.

**Comment by floam** [comment]
Except for everything curses/terminfo can't do/will never do. The structures can't accommodate 24bit colors and for some stupid reason they won't just put a darned boolean there. We could technically mitigate by adding that ourselves to every entry, and then doing a `use=actual_entry`. Still!

And all the runtime support code to detect Terminal.app/Konsole/etc, where, unlike `st`, which could be accommodated quite easily, everyelse lies about `TERM`- it can't just automagically be put in place for them transparently after we `#include` that header. We still need to do that in C++. We need to detect them correctly, pick the appropriate entries for them (the real them), in a way that won't break things for other people, or piss off users of that terminal when we haven't made some change they want in there.

That all sucks.

**Comment by krader1961** [comment]
FWIW, the [tmux project](https://github.com/tmux/tmux/) recently recognized the need to [set smkx mode](https://github.com/tmux/tmux/commit/78ca1b02656808908e2c5dc19909966c3d3bb555). Whether or not fish should do something similar is still open for debate.

**Comment by felixsanz** [comment]
Hi everyone, i've been investigating about this a few hours but i'm still confused. Whats the status on this?

If i want to get the DEL key working on `st`, what i have to do? Using `tput smkx` seems to fix it.

There is a hacky functions above that set `smkx` and `rmkx` based on events, but why is `rmkx` needed for? What if i just add `tput smkx` to my fish config file? From the last comment, it seems like some people (i.e. tmux) has `smkx` by default, with no `rmkx` anymore. I'm wrong here?

So... what is the solution here? Are we waiting for something to be implemented? This should be fixed outside at runtime?

Thanks

**Comment by wilriker** [comment]
This has been "fixed" when using `fish` inside latest `tmux` by the commit referenced by @krader1961 above - but IMHO setting `smkx` hardcoded by `tmux` is again more of a workaround then a real fix/solution.

@felixsanz `smkx` and `rmkx` are (put simply) two different modes a terminal can be in. The first is keyboard-interactive mode, i.e. the user is supposed to type in something and the latter is a mode where users are not supposed to do some inputs, e.g. because a program is still running but not expecting any input. Why pressing `DEL` in `rmkx` leads to a different keycode being interpreted crosses the borders of my historical knowledge (someone obviously had some intention when defining this). `st` is following this strict convention and if the shell (or some other process) is not setting the correct mode then `st` will interpret key presses depending on the currently set mode.

Solutions are plentiful, mostly workarounds (like setting modes based on events or as `tmux` does it by just setting `smkx` hardcoded) but the problem is doing it right, meaning implementing in it a way where it does not break compatibility with terminals that cannot handle these switches correctly.

I, personally, solved this issue by using the event-based approach plus functions that set the required mode where the event-based approach fails (namely `neovim`). This works great for me. I even extended the event-based approach a little bit to only consider `st` and interactive mode of `fish`.

```fish
if status is-interactive
    switch $TERM

        # Fix DEL key in st
        case 'st*'
            set -gx is_simple_terminal 1

        case "linux"
            set -ex is_simple_terminal
            function fish_prompt
                fish_fallback_prompt
            end
    end

    if set -q is_simple_terminal
        tput smkx ^/dev/null
        function fish_enable_keypad_transmit --on-event fish_postexec
            tput smkx ^/dev/null
        end

        function fish_disable_keypad_transmit --on-event fish_preexec
            tput rmkx ^/dev/null
        end
    end
end
```

**Comment by jdebp** [comment]
There's a lot of misinformation here.  Here's what you need to know:

* Terminals have calculator, cursor, and editing keypads.  DEC VT keyboards had the cursor keypad before the calculator keypad.  IBM PC keyboards had the calculator keypad (in the Model F) before the cursor keypad (in the Model M).
* The cursor and calculator keypads can be in either *application* or *normal* mode.  In normal mode they usually send a `DECFNK` or some other CSI-introduced control sequence.  In application mode they usually send a character shifted by SS3.  The idea was that systems that wanted essentially a big bunch of application function keys used application mode, and systems that were happy with `DECFNK` and ECMA-48-style control sequences that (say) related *both* up arrow keys to the CUU control sequence, used normal mode.
* These keypads *are* individually switchable between application and normal modes in DEC VT alikes, with the private mode settings `DECCKM` (Cursor Keypad Mode) and `DECNKM` (Numeric Keypad Mode), respectively.
* termcap and terminfo sometimes conflate actual terminal capabilities into an abstract model that is significantly limited compared to the capabilities of the actual terminals and slightly off-kilter.  In this case, `keypad_local` and `keypad_xmit` conflate the cursor and calculator keypad settings into one and imply notions of local versus remote keys, which is not in fact what they do at all.  Cursor and calculator keypad keys always transmit shift sequences and control sequences.  These capabilities merely switch between two sets.

All of this stuff *is* documented.  [Here is a list of some of the doco to read.](https://unix.stackexchange.com/a/289871/5132)

As you can see from the doco, what people are trying to emulate is this:

* The normal mode control sequence for `Delete` on the calculator keypad is either DEL or `.` according to the NumLock state, or DECFNK 3 if the ALT modifier is in effect; and the application mode shift sequence is SS3 `n`.
* The control sequence for `Delete` on the editing keypad is always DECFNK 3, because the *editing* keypad does not have an application/normal mode switch.
* The normal mode control sequence for `Left Arrow` on the calculator keypad is either CUL (i.e. CSI `D`) or `4` according to the NumLock state, or DECFNK 7 if the ALT modifier is in effect; and the application mode shift sequence is SS3 `t`.
* The normal mode control sequence for `Left Arrow` on the cursor keypad is CUL (i.e. CSI `D`), or DECFNK 7 if the ALT modifier is in effect; and the application mode shift sequence is SS3 `D`.

For what it's worth, terminfo is pretty much the wrong tool for the job for decoding this stuff.  It is wholly inadequate.  There are far more possibilities than are accountable for with a simple `kdch1=` setting.  DECFNK control sequences also incorporate modifier information as optional parameters (which is, to repeat, in the doco).

What I have done in my softwares is use an ECMA-48 state machine, checking completed CSI control sequences to see whether they are DECFNK and decoding the extended/function key and modifiers from them, decoding CSI `A` to CSI `Z` into their various DEC VT normal mode extended keys or SCO console function keys, decoding SS3 shifted characters into DEC VT application mode keys, and handling most of the rest as simple character input.

Other people take similar approaches.  NeoVIM uses [libtermkey](https://github.com/neovim/libtermkey), whose *CSI driver* is roughly a state machine too.  It does not correctly handle SCO console function keys (which is a protocol distinct from the DECFNK protocol, and conflicting with DEC's ECMA-48 CSI sequences, with a different ancestry, that you will find used by some terminal emulators including the FreeBSD kernel's built-in terminal emulator).  And it does not account for the mess that is the Linux kernel's built-in terminal emulator which erroneously swaps `Home` and `End` with `Find` and `Select`.  If you want to adopt it, you'll have to remedy those.

**Comment by faho** [comment]
The way forward here is #10359 - make the terminal report the actual keys in a better way.

I'll be closing this because keypad mode is pretty gnarly.
